xbmc/guilib/GUITextLayout.cpp

```diff
@@ -540,75 +543,101 @@ void CGUITextLayout::WrapText(const vecText &text, float maxWidth)
-
   m_lines.clear();
 
+  if (maxWidth < 0)
+  {
+    CLog::LogF(LOGWARNING, "Cannot wrap the text due to invalid max width value.");
+    return;
+  }
+
   std::vector<CGUIString> lines;
   LineBreakText(text, lines);
 
-  for (unsigned int i = 0; i < lines.size(); i++)
+  if (maxWidth == 0) // Unlimited max width
   {
-    const CGUIString &line = lines[i];
-    vecText::const_iterator lastSpace = line.m_text.begin();
+    m_lines = lines;
+    return;
+  }
+
+  size_t nMaxLines;
+  if (m_maxHeight > 0 && m_font->GetLineHeight() > 0)
+    nMaxLines = static_cast<size_t>(std::ceil(m_maxHeight / m_font->GetLineHeight()));
+  else
+    nMaxLines = std::numeric_limits<size_t>::max();
+
+  // Split lines that exceed the maximum width,
+  // lines can be split by last space char or if there are no spaces by character
+  // to mimics the line behavior of word processors
+  for (const CGUIString& line : lines)
+  {
+    if (m_lines.size() >= nMaxLines)
+      return;
+
+    if (line.m_text.empty()) // Blank line
+    {
+      m_lines.emplace_back(line);
+      continue;
+    }
+
     vecText::const_iterator pos = line.m_text.begin();
-    unsigned int lastSpaceInLine = 0;
+    vecText::const_iterator lastBeginPos = line.m_text.begin();
+    vecText::const_iterator lastSpacePos = line.m_text.end();
     vecText curLine;
-    while (pos != line.m_text.end())
+
+    while (pos < line.m_text.end())
     {
       // Get the current letter in the string
-      character_t letter = *pos;
-      // check for a space
-      if (CanWrapAtLetter(letter))
+      const character_t& letter = *pos;
+
+      if (CanWrapAtLetter(letter)) // Check for a space char
+        lastSpacePos = pos;
+
+      curLine.emplace_back(letter);
+
+      const float currWidth = m_font->GetTextWidth(curLine);
+
+      if (currWidth > maxWidth)
       {
-        float width = m_font->GetTextWidth(curLine);
-        if (width > maxWidth)
+        if (lastSpacePos > pos) // No space char where split the line, so split by char
         {
-          if (lastSpace != line.m_text.begin() && lastSpaceInLine > 0)
-          {
-            CGUIString string(curLine.begin(), curLine.begin() + lastSpaceInLine, false);
-            m_lines.push_back(string);
-            // check for exceeding our number of lines
-            if (nMaxLines > 0 && m_lines.size() >= (size_t)nMaxLines)
-              return;
-            // skip over spaces
-            pos = lastSpace;
-            while (pos != line.m_text.end() && IsSpace(*pos))
-              ++pos;
-            curLine.clear();
-            lastSpaceInLine = 0;
-            lastSpace = line.m_text.begin();
-            continue;
-          }
+          // If the pos is equal to lastBeginPos, maxWidth is not large enough to contain 1 character
+          CGUIString::iString posEnd = lastBeginPos == pos ? pos + 1 : pos;
+
+          CGUIString linePart{lastBeginPos, posEnd, false};
+          m_lines.emplace_back(linePart);
```

Comment:
Infinite loop for the maxwidth < 1st char width.

```suggestion
          // If the pos is equal to lastBeginPos, maxWidth is not large enough to contain 1 character
          // Push a line with the single character and move on to the next character.
          if (pos == lastBeginPos)
            ++pos;

            CGUIString linePart{lastBeginPos, pos, false};
            m_lines.emplace_back(linePart);
```

Reply:
ouch oversight after cleanup...thanks
proposed change is also better